base/pointer.jl

```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
 that referenced memory is not freed or garbage collected while invoking this function.
 Incorrect usage may segfault your program. Unlike C, storing memory region allocated as
 different type may be valid provided that that the types are compatible.
+
+!!! compat "Julia 1.10"
+     The `order` argument is available as of Julia 1.10.
 """
 unsafe_store!(p::Ptr{Any}, @nospecialize(x), i::Integer=1) = pointerset(p, x, Int(i), 1)
 unsafe_store!(p::Ptr{T}, x, i::Integer=1) where {T} = pointerset(p, convert(T,x), Int(i), 1)
+unsafe_store!(p::Ptr, x, order::Symbol) = atomic_pointerset(p, x, order)
+function unsafe_store!(p::Ptr, x, i::Integer, order::Symbol)
+    unsafe_store!(p + (aligned_sizeof(eltype(p)) * (Int(i) - 1)), x, order)
+end
+
+"""
+    unsafe_modify!(p::Ptr{T}, op, x, [order::Symbol]) -> Pair
+
+These atomically perform the operations to get and set a field after applying
+the function `op`.
+
+    y = unsafe_load(p)
+    z = op(y, x)
+    unsafe_store!(p, z)
+    return y => z
+
+If supported by the hardware (for example, atomic increment), this may be
+optimized to the appropriate hardware instruction, otherwise it'll use a loop.
+
+!!! compat "Julia 1.10"
+     This function requires at least Julia 1.10.
+"""
+function unsafe_modify!(p::Ptr, op, x, order::Symbol=:not_atomic)
+    Core.Intrinsics.atomic_pointermodify(p, op, x, order)
+end
+
+"""
+    unsafe_replace!(p::Ptr{T}, expected, desired,
+                  [success_order::Symbol, [fail_order::Symbol=success_order]) -> (; old, success::Bool)
+
+These atomically perform the operations to get and conditionally set a field to
+a given value.
+
+    y = unsafe_load(p, fail_order)
+    ok = y === expected
+    if ok
+        unsafe_store!(p, desired, success_order)
+    end
+    return (; old = y, success = ok)
+
+If supported by the hardware, this may be optimized to the appropriate hardware
+instruction, otherwise it'll use a loop.
+
+!!! compat "Julia 1.10"
+     This function requires at least Julia 1.10.
+"""
+function unsafe_replace!(p::Ptr{T}, expected, desired, success_order::Symbol=:not_atomic, fail_order::Symbol=success_order) where {T}
+    @inline
+    xT = desired isa T ? desired : convert(T, desired)
+    Core.Intrinsics.atomic_pointerreplace(p, expected, xT, success_order, fail_order)
+end
+function unsafe_replace!(p::Ptr{Any}, @nospecialize(expected), @nospecialize(desired), success_order::Symbol=:not_atomic, fail_order::Symbol=success_order)
+    Core.Intrinsics.atomic_pointerreplace(p, expected, desired, success_order, fail_order)
+end
+
+"""
+    unsafe_swap!(p::Ptr{T}, x, [order::Symbol])
+
+These atomically perform the operations to simultaneously get and set a field:
+
+    y = unsafe_load(p)
+    unsafe_store!(p, x)
+    return y
+
+!!! compat "Julia 1.10"
+     This function requires at least Julia 1.10.
+"""
+function unsafe_swap!(p::Ptr{Any}, x, order::Symbol=:not_atomic)
+    Core.Intrinsics.atomic_pointerswap(p, x, order)
```

Comment:
```suggestion
    return atomic_pointerswap(p, x, order)
```
(but similarly, not sure this is needed)